src/main/java/ch/njol/skript/expressions/ExprElement.java

```diff
@@ -29,87 +29,148 @@
+		int endIndex = 0;
+		if (this.startIndex != null) {
+			Number number = this.startIndex.getSingle(event);
+			if (number == null)
+				return null;
+			startIndex = number.intValue();
+		}
+		if (this.endIndex != null) {
+			Number number = this.endIndex.getSingle(event);
+			if (number == null)
+				return null;
+			endIndex = number.intValue();
+		}
 		switch (type) {
-			case FIRST:
-				element = iter.next();
+			case FIRST_ELEMENT:
+				element = iterator.next();
 				break;
-			case LAST:
-				element = Iterators.getLast(iter);
+			case LAST_ELEMENT:
+				element = Iterators.getLast(iterator);
+				break;
+			case RANDOM:
+				Object[] allElements = Iterators.toArray(iterator, Object.class);
+				element = CollectionUtils.getRandom(allElements);
 				break;
 			case ORDINAL:
-				assert this.number != null;
-				Number number = this.number.getSingle(event);
-				if (number == null)
+				if (startIndex <= 0)
 					return null;
-				try {
-					element = Iterators.get(iter, number.intValue() - 1);
-				} catch (IndexOutOfBoundsException exception) {
+				Iterators.advance(iterator, startIndex - 1);
+				if (!iterator.hasNext())
 					return null;
-				}
-				break;
-			case RANDOM:
-				Object[] allIterValues = Iterators.toArray(iter, Object.class);
-				element = CollectionUtils.getRandom(allIterValues);
+				element = iterator.next();
 				break;
 			case TAIL_END_ORDINAL:
-				allIterValues = Iterators.toArray(iter, Object.class);
-				assert this.number != null;
-				number = this.number.getSingle(event);
-				if (number == null)
-					return null;
-				int ordinal = number.intValue();
-				if (ordinal <= 0 || ordinal > allIterValues.length)
+				allElements = Iterators.toArray(iterator, Object.class);
+				if (startIndex <= 0 || startIndex > allElements.length)
 					return null;
-				element = allIterValues[allIterValues.length - ordinal];
+				element = allElements[allElements.length - startIndex];
 				break;
+			case FIRST_X_ELEMENTS:
+				if (startIndex <= 0)
+					return null;
+				//noinspection unchecked,rawtypes
+				return Iterators.toArray((Iterator) Iterators.limit(iterator, startIndex), getReturnType());
+			case LAST_X_ELEMENTS:
+				if (startIndex <= 0)
+					return null;
+				//noinspection unchecked,rawtypes
+				allElements = Iterators.toArray((Iterator) iterator, getReturnType());
+				startIndex = Math.min(startIndex, allElements.length);
+				return ArrayUtils.subarray(allElements, allElements.length - startIndex, allElements.length);
+			case RANGE:
+				//noinspection unchecked,rawtypes
+				allElements = Iterators.toArray((Iterator) iterator, getReturnType());
+				boolean reverse = startIndex > endIndex;
+				int from = Math.min(startIndex, endIndex) - 1;
+				int to = Math.max(startIndex, endIndex);
+				from = Math.max(from, 0);
+				to = Math.min(Math.max(to, 0), allElements.length);
+				Object[] elements = ArrayUtils.subarray(allElements, from, to);
+				if (reverse)
+					ArrayUtils.reverse(elements);
```

Comment:
Do we need to reverse if it's formatted as such? In this case I believe it could possibly be better to not reverse for this. Since we're not gonna care for indexes I know some people would want a gui that formats the entries in reverse so rather than

```
set {_list::*} to elements from 10 to 1 of {_list::*}
loop reversed {_list::*}
```
I believe just accepting `set {_list::*} to elements from 10 to 1 of {_list::*}` alone is a nicer mechanic
however, if not that's fine

Reply:
If you mean that `elements from 10 to 1 of {_list::*}` should return the elements from 1 to 10 but reversed, then that's what it currently does

Reply:
ah sorry, must have miss understood it, okay